Safari failed the infrastructure test unexpectedly:

```
/infrastructure/testdriver/virtual_authenticator.html
  ERROR /infrastructure/testdriver/virtual_authenticator.html - authenticator_id
Traceback (most recent call last):
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 292, in run_func
    self.result = True, self.func(self.protocol, self.url, self.timeout)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 385, in do_testharness
    done, rv = handler(result)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 676, in __call__
    return callback(url, payload)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 690, in process_action
    result = action_handler(payload)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 794, in __call__
    authenticator_id = payload["authenticator_id"]
KeyError: 'authenticator_id'
```

I am not entirely sure what's going on. I suspect this has something to do with the usage of object destructuring in https://github.com/web-platform-tests/wpt/blob/ca4e480f83ee0dbae094ec88a7a8e5851b17af93/tools/wptrunner/wptrunner/testdriver-extra.js#L138

*Perhaps* if the action is not supported, we get a null/undefined from `add_virtual_authenticator`, which then end up causing the subsequent payloads to miss the `authenticator_id` field.

We could use an explicit declaration in the front-end, or change the Python code to use `.get`: https://github.com/web-platform-tests/wpt/blob/ca4e480f83ee0dbae094ec88a7a8e5851b17af93/tools/wptrunner/wptrunner/executors/base.py#L727

--- a/infrastructure/metadata/infrastructure/testdriver/generate_test_report.html.ini
+++ b/infrastructure/metadata/infrastructure/testdriver/generate_test_report.html.ini
@@ -1,4 +1,4 @@
 [generate_test_report.html]
-  expected:
-    if product == "firefox": ERROR
-    if product == "safari" or product == "epiphany" or product == "webkit": ERROR
+  [TestDriver generate_test_report method]
+    expected:
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL

--- a/infrastructure/metadata/infrastructure/testdriver/virtual_authenticator.html.ini
+++ b/infrastructure/metadata/infrastructure/testdriver/virtual_authenticator.html.ini
@@ -1,31 +1,28 @@
 [virtual_authenticator.html]
-  expected:
-    if product == "safari" or product == "epiphany" or product == "webkit": ERROR
-
   [Can create an authenticator]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can add a credential]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can get the credentials]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can remove a credential]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can remove all credentials]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can set user verified]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL
 
   [Can remove a virtual authenticator]
     expected:
-      if product == "firefox": FAIL
+      if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": FAIL

--- a/infrastructure/metadata/infrastructure/testdriver/virtual_authenticator.html.ini
+++ b/infrastructure/metadata/infrastructure/testdriver/virtual_authenticator.html.ini
@@ -1,3 +1,31 @@
 [virtual_authenticator.html]
   expected:
-    if product == "firefox" or product == "safari" or product == "epiphany" or product == "webkit": ERROR
+    if product == "safari" or product == "epiphany" or product == "webkit": ERROR
+
+  [Can create an authenticator]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can add a credential]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can get the credentials]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can remove a credential]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can remove all credentials]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can set user verified]
+    expected:
+      if product == "firefox": FAIL
+
+  [Can remove a virtual authenticator]
+    expected:
+      if product == "firefox": FAIL

--- a/tools/wptrunner/wptrunner/executors/base.py
+++ b/tools/wptrunner/wptrunner/executors/base.py
@@ -702,7 +702,7 @@ class CallbackHandler(object):
             result = action_handler(payload)
         except NotImplementedError:
             self.logger.warning("Action %s not implemented" % action)
-            self._send_message("complete", "not-implemented")
+            self._send_message("complete", "error", "Action %s not implemented" % action)
         except Exception:
             self.logger.warning("Action %s failed" % action)
             self.logger.warning(traceback.format_exc())

--- a/tools/wptrunner/wptrunner/executors/base.py
+++ b/tools/wptrunner/wptrunner/executors/base.py
@@ -13,6 +13,7 @@ from abc import ABCMeta, abstractmethod
 
 from ..testrunner import Stop
 from .protocol import Protocol, BaseProtocolPart
+import webdriver as client
 
 here = os.path.split(__file__)[0]
 
@@ -700,7 +701,7 @@ class CallbackHandler(object):
             raise ValueError("Unknown action %s" % action)
         try:
             result = action_handler(payload)
-        except NotImplementedError:
+        except (NotImplementedError, client.UnknownCommandException):
             self.logger.warning("Action %s not implemented" % action)
             self._send_message("complete", "error", "Action %s not implemented" % action)
         except Exception:

--- a/tools/wptrunner/wptrunner/executors/base.py
+++ b/tools/wptrunner/wptrunner/executors/base.py
@@ -700,6 +700,9 @@ class CallbackHandler(object):
             raise ValueError("Unknown action %s" % action)
         try:
             result = action_handler(payload)
+        except NotImplementedError:
+            self.logger.warning("Action %s not implemented" % action)
+            self._send_message("complete", "not-implemented")
         except Exception:
             self.logger.warning("Action %s failed" % action)
             self.logger.warning(traceback.format_exc())

--- a/tools/wptrunner/wptrunner/executors/executormarionette.py
+++ b/tools/wptrunner/wptrunner/executors/executormarionette.py
@@ -31,7 +31,8 @@ from .protocol import (ActionSequenceProtocolPart,
                        ClickProtocolPart,
                        SendKeysProtocolPart,
                        TestDriverProtocolPart,
-                       CoverageProtocolPart)
+                       CoverageProtocolPart,
+                       VirtualAuthenticatorProtocolPart)
 from ..testrunner import Stop
 from ..webdriver_server import GeckoDriverServer
 
@@ -481,6 +482,30 @@ class MarionetteCoverageProtocolPart(CoverageProtocolPart):
                 # This usually happens if the process crashed
                 pass
 
+class MarionetteVirtualAuthenticatorProtocolPart(VirtualAuthenticatorProtocolPart):
+    def setup(self):
+        self.marionette = self.parent.marionette
+
+    def add_virtual_authenticator(self, config):
+        raise NotImplementedError("add_virtual_authenticator not yet implemented")
+
+    def remove_virtual_authenticator(self, authenticator_id):
+        raise NotImplementedError("remove_virtual_authenticator not yet implemented")
+
+    def add_credential(self, authenticator_id, credential):
+        raise NotImplementedError("add_credential not yet implemented")
+
+    def get_credentials(self, authenticator_id):
+        raise NotImplementedError("get_credentials not yet implemented")
+
+    def remove_credential(self, authenticator_id, credential_id):
+        raise NotImplementedError("remove_credential not yet implemented")
+
+    def remove_all_credentials(self, authenticator_id):
+        raise NotImplementedError("remove_all_credentials not yet implemented")
+
+    def set_user_verified(self, authenticator_id, uv):
+        raise NotImplementedError("set_user_verified not yet implemented")
 
 class MarionetteProtocol(Protocol):
     implements = [MarionetteBaseProtocolPart,
@@ -493,7 +518,8 @@ class MarionetteProtocol(Protocol):
                   MarionetteActionSequenceProtocolPart,
                   MarionetteTestDriverProtocolPart,
                   MarionetteAssertsProtocolPart,
-                  MarionetteCoverageProtocolPart]
+                  MarionetteCoverageProtocolPart,
+                  MarionetteVirtualAuthenticatorProtocolPart]
 
     def __init__(self, executor, browser, capabilities=None, timeout_multiplier=1, e10s=True, ccov=False):
         do_delayed_imports()

--- a/tools/wptrunner/wptrunner/executors/executormarionette.py
+++ b/tools/wptrunner/wptrunner/executors/executormarionette.py
@@ -32,6 +32,7 @@ from .protocol import (ActionSequenceProtocolPart,
                        SendKeysProtocolPart,
                        TestDriverProtocolPart,
                        CoverageProtocolPart,
+                       GenerateTestReportProtocolPart,
                        VirtualAuthenticatorProtocolPart)
 from ..testrunner import Stop
 from ..webdriver_server import GeckoDriverServer
@@ -482,6 +483,13 @@ class MarionetteCoverageProtocolPart(CoverageProtocolPart):
                 # This usually happens if the process crashed
                 pass
 
+class MarionetteGenerateTestReportProtocolPart(GenerateTestReportProtocolPart):
+    def setup(self):
+        self.marionette = self.parent.marionette
+
+    def generate_test_report(self, config):
+        raise NotImplementedError("generate_test_report not yet implemented")
+
 class MarionetteVirtualAuthenticatorProtocolPart(VirtualAuthenticatorProtocolPart):
     def setup(self):
         self.marionette = self.parent.marionette
@@ -519,6 +527,7 @@ class MarionetteProtocol(Protocol):
                   MarionetteTestDriverProtocolPart,
                   MarionetteAssertsProtocolPart,
                   MarionetteCoverageProtocolPart,
+                  MarionetteGenerateTestReportProtocolPart,
                   MarionetteVirtualAuthenticatorProtocolPart]
 
     def __init__(self, executor, browser, capabilities=None, timeout_multiplier=1, e10s=True, ccov=False):

--- a/tools/wptrunner/wptrunner/testdriver-extra.js
+++ b/tools/wptrunner/wptrunner/testdriver-extra.js
@@ -19,7 +19,7 @@
             result = JSON.parse(data.message).result
             pending_resolve(result);
         } else {
-            pending_reject();
+            pending_reject(data.status);
         }
     });
 

